arch/x86/kvm/vmx/pkvm/hyp/ptdev.c

```diff
@@ -131,5 +134,15 @@ struct pkvm_ptdev *pkvm_attach_ptdev(u16 bdf, u32 pasid, struct pkvm_shadow_vm *
 		return NULL;
 	}
 
+	/*
+	 * Reset pgt of this ptdev to VM's pgstate_pgt so need to update
+	 * IOMMU page table accordingly.
+	 */
+	ptdev->pgt = &vm->pgstate_pgt;
+	if (pkvm_iommu_sync(ptdev->bdf, ptdev->pasid)) {
+		pkvm_detach_ptdev(ptdev);
+		return NULL;
+	}
+
```

Comment:
Yes, pkvm_iommu_sync() can correct these. The reason to use cmxchg when attaching is that, it can be triggered through vmcall by the host VM. Using cmpxchg can avoid host VM maliciously using this vmcall. But this remind me that, there is probably a race between attaching and detaching. E.g., pkvm is detaching a ptdev but host VM is trying to attach this ptdev at the same time. To avoid this race, probably use cmpxchg in pkvm_detach_ptdev() after changing ptdev->pgt to host ept?

Reply:
TBH I don't quite understand how exactly the host could maliciously use the attach vmcall, could you please elaborate?

Regarding the race between attach and detach - yeah, good point, it may happen. Regarding your suggestion to use cmpxchg in pkvm_detach_ptdev() after changing ptdev->pgt to host ept, it seems it still would not prevent the situation with ptdev detached but its pgt wrongly set to pgstate_pgt? E.g.:

detach: set ptdev->pgt to host ept
attach: set ptdev->shadow_vm_handle to the VM handle
attach: set ptdev->pgt to pgstate_pgt
detach: set ptdev->shadow_vm_handle to 0

Reply:
My understanding is that host VM is not trusted, and it would be a modified linux kernel which could maliciously use the attach vmcall, right?

In the above sequence, the 2nd line "attach: set ptdev->shadow_vm_handle to the VM handle" should be failed as it is using cmpxchg and the ptdev->shadow_vm_handle is not 0, so cannot be changed to a non-zero VM id. And making attaching failed is reasonable as this attaching is not a correct behavior as this ptdev is not detached yet.

Reply:
> In the above sequence, the 2nd line "attach: set ptdev->shadow_vm_handle to the VM handle" should be failed as it is using cmpxchg and the ptdev->shadow_vm_handle is not 0, so cannot be changed to a non-zero VM id.

Ah yes indeed, forgot about that.

> My understanding is that host VM is not trusted, and it would be a modified linux kernel which could maliciously use the attach vmcall, right?

Yep, I also understand that we should be prepared for any kind of incorrect behavior of the host. I was just not getting how exactly the host could take advantage of non-atomic update of ptdev->shadow_vm_handle to achieve malicious results. Now as you pointed out that cmpxchg is actually to ensure that we will not try to attach an already attached VM, it makes sense.